### PR TITLE
accounts roles management

### PR DESCRIPTION
**Pull request by DevMike**

### CODE REVIEW

File: app/controllers/theblog/admin/accounts_controller.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
+module Theblog
+  class Admin::AccountsController < Admin::BaseItemsController
+    MODEL = Theblog::Account
+    ATTRIBUTES = %i[user_name first_name middle_name last_name email password password_confirmation]
+    INDEX = %i[user_name first_name middle_name last_name email roles]
```
Comment (kont-noor) [inline app/controllers/theblog/admin/accounts_controller.rb]: Why do you use it if you redefine `#new` `#create` etc?

File: app/models/theblog/account.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+module Theblog
+  class Account < Incarnator::Account
```
Comment (kont-noor) [inline app/models/theblog/account.rb]: I think it's better to use `Incarnator.account_model.constantize`

Comment (kont-noor) [inline app/models/theblog/account.rb]: But the model makes sense

File: app/policies/theblog/content_node_policy.rb [hunk 3]
```diff
@@ -25,7 +25,7 @@ def unblock?
     end
 
     def account_can_update_record?
-      (user == record.author && account_has_role?(user, [:editor])) ||
+      (user.try(:id) == record.author.try(:id) && account_has_role?(user, [:editor])) ||
```
Comment (kont-noor) [inline app/policies/theblog/content_node_policy.rb]: What is the reason of the change?

File: app/views/theblog/admin/_form.html.haml [hunk 4]
```diff
@@ -1,7 +1,10 @@
-= simple_form_for [:admin, item] do |f|
+= simple_form_for [:admin, item], html: { autocomplete: :off } do |f|
```
Comment (kont-noor) [inline app/views/theblog/admin/_form.html.haml]: I think it's better to use separate views and controller for this

File: spec/spec_helper.rb [hunk 5]
```diff
@@ -4,8 +4,13 @@
 require 'simplecov'
 require 'shoulda'
 require 'aasm/rspec'
+<<<<<<< 79c720011dfc0d824d291cb195114cff9edb8b4c
```
Comment (kont-noor) [inline spec/spec_helper.rb]: merging conflict

File: app/policies/theblog/content_node_policy.rb (hunk 3 above)
Comment (DevMike) [inline app/policies/theblog/content_node_policy.rb]: once I've introduced Theblog::Account, user and record author are instances of different models, Theblog::Account and Incarnator.account_model. So it's only way to compare them by their ids

File: app/views/theblog/admin/_form.html.haml (hunk 4 above)
Comment (DevMike) [inline app/views/theblog/admin/_form.html.haml]: you mean for items and accounts? As I see, the difference is not too big now, so we could keep it as I made

File: app/controllers/theblog/admin/accounts_controller.rb (hunk 1 above)
Comment (DevMike) [inline app/controllers/theblog/admin/accounts_controller.rb]: what do you mean? INDEX fields are used on the index view that is the same

File: app/policies/theblog/content_node_policy.rb (hunk 3 above)
Comment (kont-noor) [inline app/policies/theblog/content_node_policy.rb]: I think since you added an account model you also may add `Account#has_role?` method as well.
Also now we can change relation `author` to the new `Account` model

Comment (DevMike) [inline app/policies/theblog/content_node_policy.rb]: about changing relation - it was the first thing came to my mind, but once we still use Incarnator.account_model to authorise, there are a lot of other troubles.
About has_role? - could you explain?

File: app/controllers/theblog/admin/accounts_controller.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    end
+
+    private
+
+    def permitted_params
+      params.require(model_params_key).permit(*(ATTRIBUTES+[{role_ids: []}]))
+    end
+
+    def find_account
+      @account = model.find(params[:id])
```
Comment (skarm61) [inline app/controllers/theblog/admin/accounts_controller.rb]: why you don't cache item like
```
@account ||= model.find(params[:id])
```

File: app/controllers/theblog/admin/accounts_controller.rb (hunk 1 above)
Comment (skarm61) [inline app/controllers/theblog/admin/accounts_controller.rb]: How about to `.freeze` this arrays?

File: app/controllers/theblog/admin/accounts_controller.rb (hunk 6 above)
Comment (kont-noor) [inline app/controllers/theblog/admin/accounts_controller.rb]: there is the method `#item` for this

File: app/controllers/theblog/admin/accounts_controller.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    end
+
+    def item
+      @account
```
Comment (kont-noor) [inline app/controllers/theblog/admin/accounts_controller.rb]: it may be situation when `@account` is nil because it was not initialized yet.
And why you overrode the `#item` method instead of just using it?

File: app/controllers/theblog/admin/base_items_controller.rb [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    end
+    helper_method :index_fields
+
+    def items
+      @items ||= model.page params[:page]
+    end
+    helper_method :items
+
+    def item
+      raise NotImplementedError
```
Comment (kont-noor) [inline app/controllers/theblog/admin/base_items_controller.rb]: is it nesessary?

Comment (DevMike) [inline app/controllers/theblog/admin/base_items_controller.rb]: it's good practice as I know

File: app/controllers/theblog/admin/accounts_controller.rb (hunk 7 above)
Comment (DevMike) [inline app/controllers/theblog/admin/accounts_controller.rb]: the item method is implemented differently in this controller. I'll add handling for the case when account is not initialized, also I checked and there shouldn't be such case

File: spec/spec_helper.rb [hunk 9]
```diff
@@ -53,6 +61,12 @@
     mocks.verify_partial_doubles = true
   end
 
+  config.before :each do
+    Theblog::Role::PREDEFINED.each do |role| # use db cleaner
+      FactoryGirl.create(:theblog_role, name: role) unless Theblog::Role.exists?(name: role)
```
Comment (kont-noor) [inline spec/spec_helper.rb]: What about not to clear these tables?

**Comment by kont-noor** [comment]
closed because of kont-noor/Theblog#37
